tutorials/python/python-type-driven-development/tutorial.md

```diff
@@ -0,0 +1,213 @@
+
+Another concept that is important to know before we continue is the difference between weak and strong typed languages. I short, a strong typed language has stricter rules such as variable assignment, return values and function calling while weak typed ones can produce unpredictable results.
+
+That said, Python is a multi-paradigm dynamic language, so the type of variable is determined based on the value. This might be confusing, but it doesn't mean that Python has _weak typing_. Python is _strongly typed_. Surprised? Check this out:
+
+```python
+movie = "Die Hard"
+movie = movie + 2
+# TypeError: cannot concatenate 'str' and 'int' objects
+```
+
+That said, let’s begin this tutorial with a basic example:
+
+```python
+meat = "Ground beef"
+print(type(meat))
+# <class 'str'>
+
+weight_pounds = 0.5
+print(type(weight_pounds))
+# <class 'float'>
+```
+
+In the above example, we are using the `type` function to inspect the type representation of any variable. What if we include types in our first example code? Let’s check the following example:
+
+```python
+meat: str = "Ground beef"
+weight_pounds: str = 0.5
+```
+
+If we execute this file we expect the same output as the previous one, but we actually get the following:
+
+```
+error: Incompatible types in assignment (expression has type "float", variable has type "str")
+```
+
+Did you catch the reason? The variable `weight_pounds` was defined as `str` but we were assigning a float number to it. We can start spotting bugs in every type check.
+
+I’m pretty sure you can understand this function:
+
+```python
+def make_hamburger(meat, number_of_meats):
+   return ["bread"] + [meat] * number_of_meats + ["bread"]
+
+
+print(make_hamburger("ground beef", 2))
+# ['bread', 'ground beef', 'ground beef', 'bread']
+```
+
+Let’s see how it goes with types:
+
+```python
+from typing import List
+
+
+def make_hamburger(meat: str, number_of_meats: int) -> List[str]:
+   return ["bread"] + [meat] * number_of_meats + ["bread"]
+```
+
+What happened here? We defined the function’s arguments types: string for `meat`, integer for `number_of_meats` and the return type which is a list of string values. Did you see the import at the beginning? Complex types such as `List`, `Dict` or `Tuple` must be imported from the `typing` library.
+
+## Type Alias
+
+You can also define your custom type names for known structures, which is very useful for improving readability. Let’s say if we want to consider a hamburger is a list of strings, we can define a type `Hamburger` in the following way:
+
+```python
+Hamburger = List[str]
+
+def make_hamburger(meat: str, number_of_meats: int) -> Hamburger:
+   return ["bread"] + [meat] * number_of_meats + ["bread"]
+```
+
+## Callables
+
+A callable is anything you can call, using parenthesis, and possibly passing arguments. Callables are functions, classes, methods and even instances of classes can be callables if their class implements a `__call__()` method.
```

Comment:
Callables can be functions, classes, methods, or even instances of classes (if they implement a `__call__()` method).